src/resources/order/order.models.js

```diff
@@ -1,13 +1,14 @@
 class Order {
+  //[Viet Anh] not only order object from DB but also from order event from order.service.js -> limitPrice, quantity and filledQuantity must be parsed in float. Otherwise it will be in type String
```

Comment:
Hhmmmm i can't explain. These fields had been parsed to number in order controller before they get put into order service, so where do the string values come from?!!???!!
Another point is, how did they all pass the Joi schema validation while inserting into DB, in Joi schema for order collection these fields are defined to be number!

Reply:
It comes from UI when order is updated. order.service.js will parse the order object here but all number information is in type string. I can still do it manually here but I created a new order instead. So that the constructor of class Order must look like this.

Reply:
We haven't ever called  POST /order/update of order controller in our test-scenario right?, so i dont understand how these order could be ever updated?

But apart from that, the behavior is still unexplanable, because the update function of order service still executes Joi schema validation, before the updated record gets persisted into DB, so how did they manage to pass the Join schema validation??!!

Anyway... the point is, i dont want to have string values for fields which are semantically a number, and i dont want to see such input pre-processing logics like parseInt, parseFloat in service core logic and business domain models, that just violates concern seperation rule of a multi-layered architecture. We need to find out the real reason how the string value managed to get into DB.

@newmoonx: help!!!!!! :-)

Reply:
This does not come through or from DB (specially through Joi). 
If you take a look to order.service.js, when an update order event comes, new quantity and/or new limitPrice (string type) will be taken as properties of order object, the oldQuantity and oldPrice (number type) were from DB. So, these must be handled here. 
P/s: I always test by unit test and also by Postman or by Browser.

Reply:
This does not come through or from DB (specially through Joi). 
If you take a look to order.service.js, when an update order event comes, new quantity and/or new limitPrice (string type) will be taken as properties of order object, the oldQuantity and oldPrice (number type) were from DB. So, these must be handled here. 
P/s: I always test by unit test and also by Postman or by Browser.

Reply:
And of course we don't have unit test to put event through order.service (e.g. POST /order/update). It's just an issue I found when place and update order from browser

Reply:
Ok, pls fix the POST /order/update api of order controller, so that it always parse inputs/params to proper type before calling order service. And pls remove parsing logic from order service and domain models too. They don't belong there.

>This does not come through or from DB (specially through Joi).
Of course an update operation get through Joi schema validation, take a look at service.update(), Joi schema validation is called internally, right before the updated gets persisted into DB. How could these updates get through Joi schema validation is still a mystery for me.

And also remove all order records with quantity and price typed as string from your DB. They will through exception again when you want orderbook to load them while startup later.

Reply:
Yes! Problem is in order controller. We validate currently only order place event.

Reply:
https://github.com/hapijs/joi/blob/v13.6.0/API.md#number---inherits-from-any
>Generates a schema object that matches a number data type (as well as strings that can be converted to numbers). If the validation convert option is on (enabled by default), a string will be converted to a number if specified. Also, if convert is on and number.precision() is used, the value will be converted to the specified precision as well.

that explains why the updates got through Joi schema validation. We should turn off the `convert` option of Joi next time.